### ISSUE TITLE
eval dataloader 问题

### ISSUE THREAD

**Issue by ddz-mark** [PaddlePaddle/PaddleOCR#2772]
使用了 7 张图片进行 eval ，不需要进行训练。发现 acc 为 1.0，通过 infer 出来后的标签，使用 metric 求取之后 acc 为 0.85，7张图中有一张预测错误。
![image](https://user-images.githubusercontent.com/18033495/117937524-f3dde980-b2ba-11eb-9ae4-a68a0baa51cf.png)

通过日志分析之后发现， eval 预测的结果如下图所示，每个 dataloader只有 1 个样本，但是预测值中的 0 和 6 重复，1-5 重复，也就是说，一共 7 张图，只有两张图被使用，正好两张图预测正确，所以 acc 为 1.0
![image](https://user-images.githubusercontent.com/18033495/117937810-3b647580-b2bb-11eb-8711-e4bbcc26f6d1.png)

但是，通过 infer 之后发现 7 张图有 1 张预测错误，所以使用 metric 求取之后的 acc 为 0.85，想请问下为什么会出现这个情况？为什么 eval 出来的 dataloader 并没有把所有的数据使用上，而是随机取了其中的2张，重复了5张？

**Comment by ddz-mark** [comment]
配置文件 rec_chinese_lite_train_v2.0.yml 如下：

Global:
  use_gpu: True
  epoch_num: 10
  log_smooth_window: 20
  print_batch_step: 1
  save_model_dir: ./output/rec_chinese_lite_v2.0
  save_epoch_step: 3
  # evaluation is run every 5000 iterations after the 4000th iteration
  eval_batch_step: [0, 1]
  cal_metric_during_train: True
  pretrained_model:
  checkpoints: 
  save_inference_dir:
  use_visualdl: False
  infer_img: /Users/ddz/Downloads/ocrdataset/online_data/online_test_10/crop_imgs
  # for data or label process
  character_dict_path: ppocr/utils/ppocr_keys_v1.txt
  character_type: ch
  max_text_length: 25
  infer_mode: False
  use_space_char: True


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.001
  regularizer:
    name: 'L2'
    factor: 0.00001

Architecture:
  model_type: rec
  algorithm: CRNN
  Transform:
  Backbone:
    name: MobileNetV3
    scale: 0.5
    model_name: small
    small_stride: [1, 2, 2, 2]
  Neck:
    name: SequenceEncoder
    encoder_type: rnn
    hidden_size: 48
  Head:
    name: CTCHead
    fc_decay: 0.00001

Loss:
  name: CTCLoss

PostProcess:
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: SimpleDataSet
    data_dir: /Users/ddz/Downloads/ocrdataset/
    label_file_list: [ "/Users/ddz/Downloads/ocrdataset/20210420_900_train.txt" ]
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - RecAug: 
      - CTCLabelEncode: # Class handling label
      - RecResizeImg:
          image_shape: [3, 32, 320]
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: True
    batch_size_per_card: 256
    drop_last: True
    num_workers: 8

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /Users/ddz/Downloads/ocrdataset/online_data/
    label_file_list: [ "/Users/ddz/Downloads/ocrdataset/online_data/online_test_10/2021_10_test.txt" ]
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CTCLabelEncode: # Class handling label
      - RecResizeImg:
          image_shape: [3, 32, 320]
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 1
    num_workers: 0

**Comment by ddz-mark** [comment]
发现源码逻辑：ppocr/data/simple_dataset 中的 __getitem__() 方法，当 outs 为空时，eval 会取下一个 index，循环取
![image](https://user-images.githubusercontent.com/18033495/117943921-7bc6f200-b2c1-11eb-9f43-baed69214aa7.png)

**Comment by tink2123** [comment]
训练过程中需要固定输入的image_shape, 所以在train和eval时都会限制max_length. 但是预测时我们支持对任意长度文本进行预测，所以不会限制max_length.
